pycbc/detector.py

```diff
@@ -655,134 +655,810 @@ def overhead_antenna_pattern(right_ascension, declination, polarization):
+
+    offset : float (optional)
+        The time in seconds by which to offset the input waveform if
+        apply_offset is True. Default 7365189.431698299.
+    """
+    def __init__(self, reference_time=None, orbits='EqualArmlength',
+                 apply_offset=False, offset=TIME_OFFSET_20_DEGREES):
+        # specify whether to apply offsets to GPS times
+        if apply_offset:
+            self.offset = offset
+        else:
+            self.offset = 0.
+
+        # orbits properties
+        self.orbits = orbits
+        self.orbit_start_time = None
+        self.orbit_end_time = None
+
+        # waveform properties
+        self.dt = None
+        self.sample_times = None
+        if reference_time is not None:
+            self.ref_time = float(reference_time)
+        else:
+            self.ref_time = None
+        self.start_time = None
+
+        # pre- and post-processing
+        self.pad_data = False
+        self.remove_garbage = True
+        self.t0 = 1e4
+
+        # class initialization caches
+        self.proj_init = None
+        self.response_init = None
+
+    def apply_polarization(self, hp, hc, polarization):
+        """
+        Apply polarization rotation matrix.
 
-    def get_pos(self, ref_time):
-        """Return the position of LISA detector for a given reference time
         Parameters
         ----------
-        ref_time : numpy.ScalarType
+        hp : array
+            The plus polarization of the GW.
+
+        hc : array
+            The cross polarization of the GW.
+
+        polarization : float
+            The SSB polarization angle of the GW in radians.
 
         Returns
         -------
-        location : numpy.ndarray of shape (3,3)
-               Returns the position of all 3 sattelites with each row
-               correspoding to a single axis.
+        (array, array)
+            The plus and cross polarizations of the GW rotated by the
+            polarization angle.
         """
-        ref_time = Time(val=ref_time, format='gps', scale='utc').jyear
-        n = np.array(range(1, 4))
-        kappa, _lambda_ = 0, 0
-        alpha = 2. * np.pi * ref_time/1 + kappa
-        beta_n = (n - 1) * 2.0 * pi / 3.0 + _lambda_
-        a, L = 1., 0.03342293561
-        e = L/(2. * a * np.sqrt(3))
+        cphi = cos(2*polarization)
+        sphi = sin(2*polarization)
 
-        x = a*cos(alpha) + a*e*(sin(alpha)*cos(alpha)*sin(beta_n) - (1 + sin(alpha)**2)*cos(beta_n))
-        y = a*sin(alpha) + a*e*(sin(alpha)*cos(alpha)*cos(beta_n) - (1 + cos(alpha)**2)*sin(beta_n))
-        z = -np.sqrt(3)*a*e*cos(alpha - beta_n)
-        self.location = np.array([x, y, z])
+        hp_ssb = hp*cphi - hc*sphi
+        hc_ssb = hp*sphi + hc*cphi
 
-        return self.location
+        return hp_ssb, hc_ssb
 
-    def get_gcrs_pos(self, location):
-        """ Transforms ICRS frame to GCRS frame
+    def preprocess(self, hp, hc, polarization=0, reference_time=None):
```

Comment:
What does this routine mean for a truly generic implementation?

Reply:
Both the FLR and LDC versions had some common routines that needed to be applied to the waveforms before inputting them into the response generation. Mainly, this function is making sure that the signals fit into the orbital data being supplied. The original intent was to circumvent errors both backends throw when the signal extends past the orbital data, but I think this could be a generic way to ensure the waveform can be projected given the orbits.

There are some other functions in here (e.g. polarization, padding, start/ref time caching) that I added just to prevent them from being copied for each implementation, although in the interest of making the base class as generic as possible I could move these to the backend classes.